samples/native_training/main.c

```diff
@@ -0,0 +1,233 @@
+          .usage = IREE_HAL_BUFFER_USAGE_DEFAULT,
+          .type = IREE_HAL_MEMORY_TYPE_DEVICE_LOCAL,
+      },
+      iree_make_const_byte_span(state->b, sizeof(state->b)), &arg1));
+  IREE_RETURN_IF_ERROR(iree_hal_buffer_view_allocate_buffer(
+      iree_runtime_session_device_allocator(session), IREE_ARRAYSIZE(shape_X),
+      shape_X, IREE_HAL_ELEMENT_TYPE_FLOAT_32,
+      IREE_HAL_ENCODING_TYPE_DENSE_ROW_MAJOR,
+      (iree_hal_buffer_params_t){
+          .usage = IREE_HAL_BUFFER_USAGE_DEFAULT,
+          .type = IREE_HAL_MEMORY_TYPE_DEVICE_LOCAL,
+      },
+      iree_make_const_byte_span(state->X, sizeof(state->X)), &arg2));
+  IREE_RETURN_IF_ERROR(iree_hal_buffer_view_allocate_buffer(
+      iree_runtime_session_device_allocator(session), IREE_ARRAYSIZE(shape_y),
+      shape_y, IREE_HAL_ELEMENT_TYPE_FLOAT_32,
+      IREE_HAL_ENCODING_TYPE_DENSE_ROW_MAJOR,
+      (iree_hal_buffer_params_t){
+          .usage = IREE_HAL_BUFFER_USAGE_DEFAULT,
+          .type = IREE_HAL_MEMORY_TYPE_DEVICE_LOCAL,
+      },
+      iree_make_const_byte_span(state->y, sizeof(state->y)), &arg3));
+
+  // Setup call inputs with our buffers.
+  IREE_RETURN_IF_ERROR(
+      iree_runtime_call_inputs_push_back_buffer_view(&call, arg0));
+  iree_hal_buffer_view_release(arg0);
+  IREE_RETURN_IF_ERROR(
+      iree_runtime_call_inputs_push_back_buffer_view(&call, arg1));
+  iree_hal_buffer_view_release(arg1);
+  IREE_RETURN_IF_ERROR(
+      iree_runtime_call_inputs_push_back_buffer_view(&call, arg2));
+  iree_hal_buffer_view_release(arg2);
+  IREE_RETURN_IF_ERROR(
+      iree_runtime_call_inputs_push_back_buffer_view(&call, arg3));
+  iree_hal_buffer_view_release(arg3);
+
+  // Invoke the function
+  IREE_RETURN_IF_ERROR(iree_runtime_call_invoke(&call, /*flags=*/0));
+
+  // Update weights
+  iree_hal_buffer_view_t* result = NULL;
+  IREE_RETURN_IF_ERROR(
+      iree_runtime_call_outputs_pop_front_buffer_view(&call, &result));
+  IREE_RETURN_IF_ERROR(iree_hal_device_transfer_d2h(
+      iree_runtime_session_device(session), iree_hal_buffer_view_buffer(result),
+      0, &state->w, sizeof(state->w), IREE_HAL_TRANSFER_BUFFER_FLAG_DEFAULT,
+      iree_infinite_timeout()));
+
+  // Update bias
+  IREE_RETURN_IF_ERROR(
+      iree_runtime_call_outputs_pop_front_buffer_view(&call, &result));
+  IREE_RETURN_IF_ERROR(iree_hal_device_transfer_d2h(
+      iree_runtime_session_device(session), iree_hal_buffer_view_buffer(result),
+      0, &state->b, sizeof(state->b), IREE_HAL_TRANSFER_BUFFER_FLAG_DEFAULT,
+      iree_infinite_timeout()));
+
+  // Update loss
+  IREE_RETURN_IF_ERROR(
+      iree_runtime_call_outputs_pop_front_buffer_view(&call, &result));
+  IREE_RETURN_IF_ERROR(iree_hal_device_transfer_d2h(
+      iree_runtime_session_device(session), iree_hal_buffer_view_buffer(result),
+      0, &state->loss, sizeof(state->loss), IREE_HAL_TRANSFER_BUFFER_FLAG_DEFAULT,
+      iree_infinite_timeout()));
+  iree_hal_buffer_view_release(result);
```

Comment:
Note that these `iree_hal_device_transfer_d2h` calls are all synchronous. This could be rewritten to batch them together and then do a wait_all. (I don't think we have a sample showing how to do that though, and this code is easier to follow)

Reply:
Cool, yeah it seems like something we might want to do a separate walkthrough on.